File: 2024/day6/main.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"iter"
	"os"
	"strconv"
	"time"
)

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

type num = int

type grid [][]byte

func (g grid) inBounds(p pair) bool {
	return p._0 >= 0 && p._1 >= 0 && p._0 < num(len(g[0])) && p._1 < num(len(g))
}

func (g grid) Coords() iter.Seq2[pair, byte] {
	return func(yield func(pair, byte) bool) {
		for y := range len(g) {
			for x := range len(g[0]) {
				p := pair{num(x), num(y)}
				if !yield(p, g.Get(p)) {
					return
				}
			}
		}
	}
}
func (g grid) Get(p pair) byte {
	// need to access as y/x
	return g[p._1][p._0]
}
func (g grid) Set(p pair, b byte) {
	g[p._1][p._0] = b
}
func readGrid(reader *bufio.Scanner) grid {
	grid := make([][]byte, 0)
	for reader.Scan() {
		grid = append(grid, bytes.Clone(reader.Bytes()))
	}
	return grid
}

type pair struct {
	_0 num
	_1 num
}

func (p pair) String() string {
	return strconv.FormatInt(int64(p._0), 10) + "," + strconv.FormatInt(int64(p._1), 10)
}

func readMap(reader *bufio.Scanner) (grid, pair) {
	grid := readGrid(reader)
	start := pair{}
	for p, b := range grid.Coords() {
		if b == '^' {
			start = p
			break
		}
	}
	return grid, start
}

type dir int

const (
	UP dir = iota
	RIGHT
	DOWN
	LEFT
)

func turn90(dir dir) dir {
	return (dir + 1) % 4
}

func (p pair) move(dir dir) pair {
	// top left is 0,0 so right is x increasing and down is y increasing
	newp := pair{p._0, p._1}
	switch dir {
	case UP:
		newp._1 -= 1
	case DOWN:
		newp._1 += 1
	case RIGHT:
		newp._0 += 1
	case LEFT:
		newp._0 -= 1
	}
	return newp
}

func fullPath(grid grid, start pair) []locdir {
	seenDir := make([]locdir, 0)
	p := start
	currentDir := UP
	for {
		seenDir = append(seenDir, locdir{p, currentDir})
		nextp := p.move(currentDir)
		if !grid.inBounds(nextp) {
			break
		}
		for grid.Get(nextp) == '#' { // doesn't handle nextp being out of bounds technically
			currentDir = turn90(currentDir)
			nextp = p.move(currentDir)
		}
		p = nextp
	}
	return seenDir
}

func solve1(reader *bufio.Scanner) {
	grid, start := readMap(reader)
	seenDir := fullPath(grid, start)
	seen := make(map[pair]struct{})
	for _, p := range seenDir {
		seen[p.p] = struct{}{}
	}
	fmt.Println(len(seen))
}

type locdir struct {
	p pair
	d dir
}

func causeLoop(grid grid, start locdir, block pair) bool {
	// seen := make(map[pair]struct{})
	seenDir := make(map[locdir]struct{})
	p := start.p
	currentDir := start.d
	for {
		// seen[p] = struct{}{}
		curLocDir := locdir{p, currentDir}
		if _, ok := seenDir[curLocDir]; ok { // been here in current direction, i.e. loop
			return true
		}
		seenDir[curLocDir] = struct{}{}
		nextp := p.move(currentDir)
		if !grid.inBounds(nextp) {
			return false
		}

		for grid.Get(nextp) == '#' || nextp == block {
			currentDir = turn90(currentDir)
			nextp = p.move(currentDir)
		}
		p = nextp
	}
}

func solve2(reader *bufio.Scanner) {
	t1 := time.Now()
	grid, start := readMap(reader)
	t2 := time.Now()
	seenDir := fullPath(grid, start)
	t3 := time.Now()
	count := 0
	seen := make(map[pair]struct{})
	seen[start] = struct{}{}
	for i := 1; i < len(seenDir); i++ {
		prevp := seenDir[i-1]
		p := seenDir[i]
		if _, ok := seen[p.p]; ok { // already checked, handles start
			continue
		}
		seen[p.p] = struct{}{}
		if causeLoop(grid, prevp, p.p) {
			count++
		}
	}
	t4 := time.Now()
	fmt.Println(t2.Sub(t1), t3.Sub(t2), t4.Sub(t3))
	fmt.Println(count)
}

func main() {
	f, err := os.Open(os.Args[1])
	defer f.Close()
	if err != nil {
		fmt.Printf("Error opening file: %v", err)
		return
	}
	reader := bufio.NewReader(f)
	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 130*131+10), bufio.MaxScanTokenSize)
	// solve1(scanner) // 41 - 5409
	solve2(scanner) // 6 - 2022
}

File: 2024/day6/main.zig
const std = @import("std");
// const input = @embedFile("example.txt");
const input = @embedFile("input.txt");

var gpa = std.heap.GeneralPurposeAllocator(.{}){};
const allocator = gpa.allocator();

const Grid = struct {
    g: std.ArrayList([]const u8),

    fn inBounds(self: Grid, p: Point) bool {
        return p.x >= 0 and p.y >= 0 and p.x < self.g.items[0].len and p.y < self.g.items.len;
    }

    fn Get(self: Grid, p: Point) u8 {
        return self.g.items[@intCast(p.y)][@intCast(p.x)];
    }
};

const Point = struct {
    x: i64,
    y: i64,

    fn move(self: Point, dir: Direction) Point {
        var x = self.x;
        var y = self.y;
        switch (dir) {
            Direction.up => y -= 1,
            Direction.down => y += 1,
            Direction.right => x += 1,
            Direction.left => x -= 1,
        }
        return Point{ .x = x, .y = y };
    }
};

const PointDir = struct {
    p: Point,
    d: Direction,
};

const Direction = enum(u2) {
    up,
    right,
    down,
    left,
    fn turn90(self: Direction) Direction {
        const res = @addWithOverflow(@intFromEnum(self), 1);
        return @enumFromInt(res[0]);
    }
};

const None = struct {};

var start = Point{ .x = 0, .y = 0 };

fn readGrid() !Grid {
    var it = std.mem.tokenizeScalar(u8, input, '\n');
    var g = Grid{
        .g = std.ArrayList([]const u8).init(allocator),
    };
    var y: i64 = 0;
    while (it.next()) |token| {
        try g.g.append(token);
        for (token, 0..) |b, x| {
            if (b == '^') {
                start = Point{ .x = @intCast(x), .y = y };
            }
        }
        y += 1;
    }
    return g;
}

fn findStart(g: Grid) ?Point {
    for (g.g.items, 0..) |row, y| {
        for (row, 0..) |b, x| {
            if (b == '^') {
                return Point{ .x = @intCast(x), .y = @intCast(y) };
            }
        }
    }
    return null;
}

fn fullPath(g: Grid) ![]PointDir {
    var seenDir = std.ArrayList(PointDir).init(allocator);
    var p = start;
    var currentDir: Direction = Direction.up;
    while (true) {
        try seenDir.append(PointDir{ .p = p, .d = currentDir });
        var nextp = p.move(currentDir);
        if (!g.inBounds(nextp)) {
            break;
        }
        while (g.Get(nextp) == '#') { // doesn't handle nextp being out of bounds technically
            currentDir = currentDir.turn90();
            nextp = p.move(currentDir);
        }
        p = nextp;
    }
    return seenDir.items;
}

fn causeLoop(grid: Grid, startP: PointDir, block: Point) !bool {
    var seenDir = std.AutoHashMap(PointDir, void).init(allocator);
    var p = startP.p;
    var currentDir = startP.d;
    while (true) {
        const curLocDir = PointDir{ .p = p, .d = currentDir };
        if (seenDir.contains(curLocDir)) { // been here in current direction, i.e. loop
            return true;
        }
        try seenDir.put(curLocDir, {});
        var nextp = p.move(currentDir);
        if (!grid.inBounds(nextp)) {
            return false;
        }

        while (grid.Get(nextp) == '#' or (nextp.x == block.x and nextp.y == block.y)) {
            currentDir = currentDir.turn90();
            nextp = p.move(currentDir);
        }
        p = nextp;
    }
}

fn solve1(grid: Grid) !void {
    const path = try fullPath(grid);
    var map = std.AutoHashMap(Point, void).init(allocator);
    for (path) |p| {
        try map.put(p.p, {});
    }
    std.debug.print("{d}\n", .{map.count()});
}

fn solve2(grid: Grid) !void {
    const seenDir = try fullPath(grid);
    var count: u32 = 0;
    var seen = std.AutoHashMap(Point, void).init(allocator);
    try seen.put(start, {});
    for (seenDir[0 .. seenDir.len - 1], seenDir[1..seenDir.len]) |prevp, p| {
        if (seen.contains(p.p)) { // already checked, handles start
            continue;
        }
        try seen.put(p.p, {});
        if (try causeLoop(grid, prevp, p.p)) {
            count = count + 1;
        }
    }
    std.debug.print("{d}\n", .{count});
}

pub fn main() !void {
    const grid = try readGrid();
    // try solve1(grid);
    try solve2(grid);
}
